File: programers_Lv2/타겟 넘버.js
function solution(numbers, target) {
  var answer = 0;
  const sumNumbers = numbers.reduce((a, b) => a + b);

  function getSubArr(arr) {
    const subArr = [];

    function backTrack(start, curr) {
      subArr.push([...curr]);

      for (let i = start; i < arr.length; i++) {
        curr.push(arr[i]);
        backTrack(i + 1, curr);
        curr.pop();
      }
    }

    backTrack(0, []);
    return subArr;
  }

  const tempArr = getSubArr(numbers);

  tempArr.forEach((v, i) => {
    if (v.length !== 0) {
      const sumArr = v.reduce((a, b) => a + b);
      answer += sumNumbers - sumArr * 2 === target ? 1 : 0;
    }
  });

  return answer;
}

console.log(solution([1, 1, 1, 1, 1], 3)); // 5
console.log(solution([4, 1, 2, 1], 4)); // 2

File: programers_Lv2/타겟 넘버.py
# https://school.programmers.co.kr/learn/courses/30/lessons/43165
def solution(numbers, target):
    answer = 0
    sum_numbers = sum(numbers)

    def get_sublists(arr):
        sublists = []

        def backtrack(start, curr):
            sublists.append(tuple(curr))

            for i in range(start, len(arr)):
                curr.append(arr[i])
                backtrack(i + 1, curr)
                curr.pop()

        backtrack(0, [])
        return sublists

    temp_lst = get_sublists(numbers)

    for t in temp_lst:
        if sum_numbers - (sum(t) * 2) == target:
            answer += 1

    return answer


print(solution([1, 1, 1, 1, 1], 3))  # 5
print(solution([4, 1, 2, 1], 4))  # 2
